rpc/src/account_history.rs

```diff
@@ -0,0 +1,226 @@
+    thread_hdl: JoinHandle<()>,
+}
+
+impl RpcAccountHistoryService {
+    pub fn new(
+        config: &AccountHistoryConfig,
+        account_keys: Arc<RwLock<AccountKeys>>,
+        account_history: Arc<RwLock<AccountHistory>>,
+        account_history_receiver: Receiver<Arc<Bank>>,
+        exit: &Arc<AtomicBool>,
+    ) -> Self {
+        let exit = exit.clone();
+        let config = config.clone();
+        let thread_hdl = Builder::new()
+            .name("solana-account-history".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+                if let Err(RecvTimeoutError::Disconnected) = Self::receive_bank(
+                    &config,
+                    &account_keys,
+                    &account_history,
+                    &account_history_receiver,
+                ) {
+                    break;
+                }
+            })
+            .unwrap();
+        Self { thread_hdl }
+    }
+
+    fn receive_bank(
+        config: &AccountHistoryConfig,
+        account_keys: &Arc<RwLock<AccountKeys>>,
+        account_history: &Arc<RwLock<AccountHistory>>,
+        account_history_receiver: &Receiver<Arc<Bank>>,
+    ) -> Result<(), RecvTimeoutError> {
+        let frozen_bank = account_history_receiver.recv_timeout(Duration::from_secs(1))?;
+        debug!(
+            "rpc account-history service received bank: {:?}",
+            frozen_bank.slot()
+        );
+
+        let r_account_keys = account_keys.read().unwrap();
+        let mut measure_collect = Measure::start("collect-account-history");
+        let slot_accounts =
+            Self::collect_accounts(&frozen_bank, &r_account_keys, config.updates_only);
+        measure_collect.stop();
+        drop(r_account_keys);
+
+        let mut measure_write = Measure::start("write-account-history");
+        let mut w_account_history = account_history.write().unwrap();
+        w_account_history.insert(frozen_bank.slot(), slot_accounts);
+        measure_write.stop();
+
+        let mut measure_prune = Measure::start("prune-account-history");
+        Self::remove_old_slots(w_account_history, &config.num_slots);
+        measure_prune.stop();
+
+        datapoint_info!(
+            "rpc-account-history",
+            ("collect", measure_collect.as_us(), i64),
+            ("write", measure_write.as_us(), i64),
+            ("prune", measure_prune.as_us(), i64),
+        );
+
+        Ok(())
+    }
+
+    fn collect_accounts(
+        frozen_bank: &Bank,
+        accounts: &HashSet<Pubkey>,
+        updates_only: bool,
+    ) -> HashMap<Pubkey, Account> {
+        let mut slot_accounts = HashMap::new();
+        for address in accounts.iter() {
+            let shared_account = if updates_only {
+                frozen_bank
+                    .get_account_modified_slot(address)
+                    .filter(|(_shared_account, slot)| *slot == frozen_bank.slot())
```

Comment:
hmmm if the intention is to capture *exactly* the update for this account, at `frozen_bank.slot()`, this may race with account clean if we make a root past this `frozen_bank.slot()` before this call to `get_account_modified_slot` runs.

Reply:
Makes sense. I'm a little reluctant to dangle bits into AccountsBackgroundService for this, unless some sort of minimum-root-to-keep tracker would be useful for other purposes?
Which approach do you prefer?

Reply:
I think 2. is probably the easiest :), but also open to suggestions from @sakridge or @ryoqun

Reply:
Since we both lean toward solution 2, and having heard no feedback from those other two, I will start moving on that.

Reply:
👻